File: kotlin-root/gasnake/src/main/kotlin/snake/game/Snake.kt
package com.example.snake.game

import com.example.snake.game.typing.Direction
import com.example.snake.game.typing.NotNullPositionAndDirection
import com.example.snake.game.typing.PositionAndDirection
import com.example.snake.game.typing.SnakeAction

class Snake(
    positions: MutableList<Position>,
    var direction: Direction,
    val allPositions2D: List<List<Position>>
) {
    companion object {
        fun actionMap(direction: Direction, snakeAction: SnakeAction): DirectionAndDxDy = when (direction) {
            Direction.UP -> when (snakeAction) {
                SnakeAction.FRONT -> DirectionAndDxDy.UP
                SnakeAction.TURN_LEFT -> DirectionAndDxDy.LEFT
                SnakeAction.TURN_RIGHT -> DirectionAndDxDy.RIGHT
            }

            Direction.DOWN -> when (snakeAction) {
                SnakeAction.FRONT -> DirectionAndDxDy.DOWN
                SnakeAction.TURN_LEFT -> DirectionAndDxDy.RIGHT
                SnakeAction.TURN_RIGHT -> DirectionAndDxDy.LEFT
            }

            Direction.LEFT -> when (snakeAction) {
                SnakeAction.FRONT -> DirectionAndDxDy.LEFT
                SnakeAction.TURN_LEFT -> DirectionAndDxDy.DOWN
                SnakeAction.TURN_RIGHT -> DirectionAndDxDy.UP
            }

            Direction.RIGHT -> when (snakeAction) {
                SnakeAction.FRONT -> DirectionAndDxDy.RIGHT
                SnakeAction.TURN_LEFT -> DirectionAndDxDy.UP
                SnakeAction.TURN_RIGHT -> DirectionAndDxDy.DOWN
            }
        }

        fun directionDxDyMap(direction: Direction): DxDy = when (direction) {
            Direction.UP -> DxDy.UP
            Direction.DOWN -> DxDy.DOWN
            Direction.LEFT -> DxDy.LEFT
            Direction.RIGHT -> DxDy.RIGHT
        }
    }

    val positions = ArrayDeque(positions)
    val positionsSet = positions.toMutableSet()
    val head: Position
        get() = positions[0]
    val length: Int
        get() = positions.size


    fun positionIsInSnake(position: Position): Boolean = positionsSet.contains(position)

    fun toPlainObject(): SnakeData = SnakeData(positions, direction, allPositions2D)

    fun toPlainObjectWithoutAllPositions2D(): SnakeData = SnakeData(positions, direction, emptyList())

    /**
     * Checks if the snake will eat itself after moving to a new position.
     *
     * @param position The new position of the head (assumes this position is not the food position).
     * @return True if the new position is occupied by the snake, false otherwise.
     */
    fun checkEatSelfAfterMove(position: Position): Boolean {
        // Ignore the tail as it will move
        val lastIndex = this.positions.size - 1
        return if (this.positions[lastIndex] == position) false else positionIsInSnake(position)
    }

    fun getHeadPositionAndDirectionAfterMoveBySnakeAction(action: SnakeAction): PositionAndDirection {
        val (x0, y0) = head
        val (direction, positionChange) = actionMap(direction, action)
        val (dx, dy) = positionChange
        val position: Position? = allPositions2D.getOrNull(y0 + dy)?.getOrNull(x0 + dx)
        return PositionAndDirection(position, direction)
    }

    fun getHeadPositionAndDirectionAfterMoveByDirection(direction: Direction): PositionAndDirection {
        val (x0, y0) = head
        val (dx, dy) = directionDxDyMap(direction)
        val position: Position? = allPositions2D.getOrNull(y0 + dy)?.getOrNull(x0 + dx)
        return PositionAndDirection(position, direction)
    }

    fun move(direction: Direction) {
        val positionAndDirection =
            getHeadPositionAndDirectionAfterMoveByDirectionWithCheckingEatSelfAndOutOfBound(direction)

        positionsSet.remove(positions.last())
        positionsSet.add(positionAndDirection.position)

        positions.addFirst(positionAndDirection.position)
        positions.removeLast()

        this.direction = positionAndDirection.direction
    }

    fun moveWithFoodEaten(direction: Direction) {
        val positionAndDirection =
            getHeadPositionAndDirectionAfterMoveByDirectionWithCheckingEatSelfAndOutOfBound(direction)
        positionsSet.add(positionAndDirection.position)
        positions.addFirst(positionAndDirection.position)
        this.direction = positionAndDirection.direction
    }

    private fun getHeadPositionAndDirectionAfterMoveByDirectionWithCheckingEatSelfAndOutOfBound(direction: Direction): NotNullPositionAndDirection {
        if (this.direction == direction.opposite) throw IllegalArgumentException("Snake cannot move to opposite direction")
        val positionAndDirection = getHeadPositionAndDirectionAfterMoveByDirection(direction)
        positionAndDirection.position ?: throw IllegalArgumentException("Snake move will be out of bounds")
        return NotNullPositionAndDirection(positionAndDirection.position, positionAndDirection.direction)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Snake

        if (direction != other.direction) return false
        if (allPositions2D != other.allPositions2D) return false
        if (positions != other.positions) return false

        return true
    }

    override fun hashCode(): Int {
        var result = direction.hashCode()
        result = 31 * result + allPositions2D.hashCode()
        result = 31 * result + positions.hashCode()
        return result
    }
}

data class DxDy(val dx: Int, val dy: Int) {
    companion object {
        val UP = DxDy(0, -1)
        val DOWN = DxDy(0, 1)
        val LEFT = DxDy(-1, 0)
        val RIGHT = DxDy(1, 0)
    }
}

data class DirectionAndDxDy(val direction: Direction, val dxDy: DxDy) {
    companion object {
        val UP = DirectionAndDxDy(Direction.UP, DxDy.UP)
        val DOWN = DirectionAndDxDy(Direction.DOWN, DxDy.DOWN)
        val LEFT = DirectionAndDxDy(Direction.LEFT, DxDy.LEFT)
        val RIGHT = DirectionAndDxDy(Direction.RIGHT, DxDy.RIGHT)
    }
}

data class SnakeData(
    val positions: List<Position>,
    val direction: Direction,
    val allPositions2D: List<List<Position>>
)

File: node-root/snake-game/Snake.ts
import { Direction, SnakeAction } from "./typing";
import { oppositeDirection } from "./oppositeDirection";
import type Position from "./Position";
import type { IPosition } from "./Position";
import type { PositionAndDirection, NotNullPositionAndDirection } from "./typing";

interface ActionMapSecondLayerObject {
  direction: Direction;
  positionChange: { dx: number; dy: number };
}

interface ActionMapFirstLayerObject {
  [SnakeAction.FRONT]: Readonly<ActionMapSecondLayerObject>;
  [SnakeAction.TURN_LEFT]: Readonly<ActionMapSecondLayerObject>;
  [SnakeAction.TURN_RIGHT]: Readonly<ActionMapSecondLayerObject>;
}

interface ActionMap {
  [Direction.UP]: Readonly<ActionMapFirstLayerObject>;
  [Direction.DOWN]: Readonly<ActionMapFirstLayerObject>;
  [Direction.LEFT]: Readonly<ActionMapFirstLayerObject>;
  [Direction.RIGHT]: Readonly<ActionMapFirstLayerObject>;
}

interface DirectionDxDyMap {
  [Direction.UP]: { dx: number; dy: number };
  [Direction.DOWN]: { dx: number; dy: number };
  [Direction.LEFT]: { dx: number; dy: number };
  [Direction.RIGHT]: { dx: number; dy: number };
}

export interface ISnake {
  positions: IPosition[];
  direction: Direction;
  allPositions2D: IPosition[][];
}

export default class Snake implements ISnake {
  private static actionMap: ActionMap = {
    [Direction.UP]: {
      [SnakeAction.FRONT]: { direction: Direction.UP, positionChange: { dx: 0, dy: -1 } },
      [SnakeAction.TURN_LEFT]: { direction: Direction.LEFT, positionChange: { dx: -1, dy: 0 } },
      [SnakeAction.TURN_RIGHT]: { direction: Direction.RIGHT, positionChange: { dx: 1, dy: 0 } },
    },
    [Direction.DOWN]: {
      [SnakeAction.FRONT]: { direction: Direction.DOWN, positionChange: { dx: 0, dy: 1 } },
      [SnakeAction.TURN_LEFT]: { direction: Direction.RIGHT, positionChange: { dx: 1, dy: 0 } },
      [SnakeAction.TURN_RIGHT]: { direction: Direction.LEFT, positionChange: { dx: -1, dy: 0 } },
    },
    [Direction.LEFT]: {
      [SnakeAction.FRONT]: { direction: Direction.LEFT, positionChange: { dx: -1, dy: 0 } },
      [SnakeAction.TURN_LEFT]: { direction: Direction.DOWN, positionChange: { dx: 0, dy: 1 } },
      [SnakeAction.TURN_RIGHT]: { direction: Direction.UP, positionChange: { dx: 0, dy: -1 } },
    },
    [Direction.RIGHT]: {
      [SnakeAction.FRONT]: { direction: Direction.RIGHT, positionChange: { dx: 1, dy: 0 } },
      [SnakeAction.TURN_LEFT]: { direction: Direction.UP, positionChange: { dx: 0, dy: -1 } },
      [SnakeAction.TURN_RIGHT]: { direction: Direction.DOWN, positionChange: { dx: 0, dy: 1 } },
    },
  };

  private static readonly directionDxDyMap: DirectionDxDyMap = {
    [Direction.UP]: { dx: 0, dy: -1 },
    [Direction.DOWN]: { dx: 0, dy: 1 },
    [Direction.LEFT]: { dx: -1, dy: 0 },
    [Direction.RIGHT]: { dx: 1, dy: 0 },
  };

  public direction: Direction;
  public readonly positions: Position[];
  public readonly allPositions2D: Position[][];

  constructor(positions: Position[], direction: Direction, allPositions2D: Position[][]) {
    this.positions = positions.slice();
    this.direction = direction;
    this.allPositions2D = allPositions2D;
  }

  public toPlainObject(): ISnake {
    return {
      positions: this.positions.map((position) => position.toPlainObject()),
      direction: this.direction,
      allPositions2D: this.allPositions2D.map((row) => row.map((position) => position.toPlainObject())),
    };
  }

  public toPlainObjectWithoutAllPositions2D(): ISnake {
    return {
      positions: this.positions.map((position) => position.toPlainObject()),
      direction: this.direction,
      allPositions2D: [],
    };
  }

  public get head(): Position {
    return this.positions[0];
  }

  public get tail(): Position {
    return this.positions[this.length - 1];
  }

  public get length(): number {
    return this.positions.length;
  }

  public positionInSnake(position: Position): boolean {
    return this.positions.some((x) => x.isEqual(position));
  }

  /**
   *
   * @param position new position of the head (assume this position is not the food position)
   * @returns true if the new position is occupied by the snake
   */
  public checkEatSelfAfterMove(position: Position): boolean {
    // ignore the tail as it will move
    return this.tail.isEqual(position) ? false : this.positionInSnake(position);
  }

  /**
   * return null position if snake will be out of bounds
   */
  public getHeadPositionAndDirectionAfterMoveBySnakeAction(action: SnakeAction): PositionAndDirection {
    const { x: x0, y: y0 } = this.head;
    const { direction, positionChange } = Snake.actionMap[this.direction][action];
    const { dx, dy } = positionChange;
    const position: Position | null = this.allPositions2D[y0 + dy]?.[x0 + dx] ?? null;
    return { position, direction };
  }

  public getHeadPositionAndDirectionAfterMoveByDirection(direction: Direction): PositionAndDirection {
    const { x: x0, y: y0 } = this.head;
    const { dx, dy } = Snake.directionDxDyMap[direction];
    const position: Position | null = this.allPositions2D[y0 + dy]?.[x0 + dx] ?? null;
    return { position, direction };
  }

  public move(direction: Direction): void {
    const positionAndDirection = this.getHeadPositionAndDirectionAfterMoveByDirectionWithCheckingEatSelfAndOutOfBound(direction);
    this.positions.unshift(positionAndDirection.position);
    this.positions.pop();
    this.direction = positionAndDirection.direction;
  }

  public moveWithFoodEaten(direction: Direction): void {
    const positionAndDirection = this.getHeadPositionAndDirectionAfterMoveByDirectionWithCheckingEatSelfAndOutOfBound(direction);
    this.positions.unshift(positionAndDirection.position);
    this.direction = positionAndDirection.direction;
  }

  private getHeadPositionAndDirectionAfterMoveByDirectionWithCheckingEatSelfAndOutOfBound(direction: Direction): NotNullPositionAndDirection {
    if (this.direction === oppositeDirection(direction)) throw new Error("snake cannot move to opposite direction");
    const positionAndDirection = this.getHeadPositionAndDirectionAfterMoveByDirection(direction);
    if (positionAndDirection.position === null) throw new Error("snake move will out of bound");
    return {
      position: positionAndDirection.position,
      direction: positionAndDirection.direction,
    };
  }
}
